include/curl/curl.h

```diff
@@ -640,6 +640,9 @@ typedef enum {
   CURLE_PROXY,                   /* 97 - proxy handshake error */
   CURLE_SSL_CLIENTCERT,          /* 98 - client-side certificate required */
   CURLE_UNRECOVERABLE_POLL,      /* 99 - poll/select returned fatal error */
+#ifdef USE_ECH
+  CURLE_ECH_REQUIRED      ,      /* 100 - ECH tried but failed */
```

Comment:
```suggestion
  CURLE_ECH_FAILED      ,      /* 100 - ECH tried but failed */
```

Mostly because this makes the error code spell out what actually happened.

Reply:
I need to look at this some more. The ECH spec defines the new "ech_required" alert that clients use to close the connection in this case, which is why I used that. OTOH, boringssl does differentiate between that and an "ech failed" API error code, so there's a possibly silly divergence between my OpenSSL fork and boringssl there. As @davidben is smarter than me he's probably correct there, so that's one I'll revisit, but not right now.

So, I've left this as-is for the moment.

Reply:
I had a look at this. While I could make a change like that, I reckon it's likely better to leave this as-is for now on the basis that once there's a PR submitted to the OpenSSL maintainers, I'm pretty sure they'll want changes in OpenSSL error codes, maybe that one, but probably others as well, and those changes would likely percolate up to the curl code too, so a change now probably isn't worth it for now.

Reply:
"Required" was named that way because failure to decrypt in ECH isn't immediately fatal to the protocol. You just handshake with ClientHelloOuter instead. That's what allows ECH GREASE to work. Of course, the client has an opinion as to which of ClientHelloOuter or ClientHelloInner it wants, but the server doesn't know that. So the intended read was "hey, you (server) didn't decrypt ECH and I (client) required it to accept this connection".